app/src/main/java/com/hedvig/app/feature/embark/passages/addressautocomplete/EmbarkAddressAutoCompleteFragment.kt

```diff
@@ -0,0 +1,149 @@
+import com.hedvig.app.feature.embark.EmbarkViewModel
+import com.hedvig.app.feature.embark.Response
+import com.hedvig.app.feature.embark.passages.MessageAdapter
+import com.hedvig.app.feature.embark.passages.addressautocomplete.composables.AddressCard
+import com.hedvig.app.feature.embark.passages.animateResponse
+import com.hedvig.app.feature.embark.ui.EmbarkActivity
+import com.hedvig.app.ui.compose.theme.HedvigTheme
+import com.hedvig.app.util.extensions.view.applyNavigationBarInsets
+import com.hedvig.app.util.extensions.view.hapticClicks
+import com.hedvig.app.util.extensions.viewLifecycle
+import com.hedvig.app.util.extensions.viewLifecycleScope
+import com.zhuinden.fragmentviewbindingdelegatekt.viewBinding
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.mapLatest
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+import org.koin.androidx.viewmodel.ext.android.sharedViewModel
+import org.koin.androidx.viewmodel.ext.android.viewModel
+import org.koin.core.parameter.parametersOf
+
+class EmbarkAddressAutoCompleteFragment : Fragment(R.layout.fragment_embark_address_auto_complete_action) {
+
+    private val data: EmbarkAddressAutoCompleteParams
+        get() = requireArguments().getParcelable(DATA)
+            ?: throw Error("Programmer error: DATA is null in ${this.javaClass.name}")
+
+    private val embarkViewModel: EmbarkViewModel by sharedViewModel()
+    private val viewModel: EmbarkAddressAutoCompleteViewModel by viewModel {
+        val prefilledAddress: DanishAddress? = DanishAddress.fromValueStoreKeys(embarkViewModel::getPrefillFromStore)
+        parametersOf(prefilledAddress)
+    }
+    private val binding by viewBinding(FragmentEmbarkAddressAutoCompleteActionBinding::bind)
+
+    private lateinit var fetchDanishAddressAutoCompleteContractHandler: FetchDanishAddressAutoCompleteContractHandler
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        fetchDanishAddressAutoCompleteContractHandler = FetchDanishAddressAutoCompleteContractHandler(
+            registry = requireActivity().activityResultRegistry,
+            onAddressResult = { result ->
+                when (result) {
+                    FetchDanishAddressContractResult.CantFind -> {
+                        lifecycleScope.launch {
+                            saveAndAnimate(null)
+                            embarkViewModel.submitAction(data.link)
+                        }
+                    }
+                    is FetchDanishAddressContractResult.Selected -> {
+                        viewModel.updateAddressSelected(result.address)
+                    }
+                    else -> {
+                        // no-op
+                    }
+                }
+            }
+        )
+        lifecycle.addObserver(fetchDanishAddressAutoCompleteContractHandler)
```

Comment:
This is all that's required on the call site to setup the activity result, along with starting the process in line 101:103.

Reply:
So good